transaction/std/src/transaction_builder.rs

```diff
@@ -280,12 +284,205 @@ impl<FPR: FogPubkeyResolver> TransactionBuilder<FPR> {
+            return Err(SignedContingentInputError::BlockVersionMismatch(
+                *self.block_version,
+                sci.block_version,
+            ));
+        }
+        if !self.block_version.masked_amount_v2_is_supported() {
+            return Err(
+                SignedContingentInputError::FeatureNotSupportedAtBlockVersion(
+                    *self.block_version,
+                    "partial fills",
+                ),
+            );
+        }
+        // Check if the sci already has membership proofs, the caller is supposed to do
+        // that
+        if sci.tx_in.ring.len() != sci.tx_in.proofs.len() {
+            return Err(SignedContingentInputError::MissingProofs);
+        }
+        let rules = sci
+            .tx_in
+            .input_rules
+            .as_ref()
+            .ok_or(SignedContingentInputError::MissingRules)?;
+        let fractional_change = rules
+            .fractional_change
+            .as_ref()
+            .ok_or(SignedContingentInputError::MissingFractionalChange)?;
+        let fractional_change_amount = fractional_change.reveal_amount()?;
+        if fractional_change_amount.value == 0 {
+            return Err(SignedContingentInputError::ZeroFractionalChange);
+        }
+        if fractional_change_amount.token_id != sci_change_amount.token_id {
+            return Err(SignedContingentInputError::TokenIdMismatch);
+        }
+        if fractional_change_amount.value < sci_change_amount.value {
+            return Err(SignedContingentInputError::ChangeExceededOffer);
+        }
+        if rules.max_allowed_change_value != 0
+            && rules.max_allowed_change_value < sci_change_amount.value
+        {
+            return Err(SignedContingentInputError::ChangeLimitExceeded);
+        }
+
+        let fill_fraction_num = (fractional_change_amount.value - sci_change_amount.value) as u128;
+        let fill_fraction_denom = fractional_change_amount.value as u128;
+
+        // Ensure that we can reveal all amounts
+        let fractional_outputs_and_amounts: Vec<(&RevealedTxOut, Amount)> = rules
+            .fractional_outputs
+            .iter()
+            .map(
+                |r_tx_out| -> Result<(&RevealedTxOut, Amount), RevealedTxOutError> {
+                    let amount = r_tx_out.reveal_amount()?;
+                    Ok((r_tx_out, amount))
+                },
+            )
+            .collect::<Result<_, _>>()?;
+
+        // Fill all fractional outputs to precisely the smallest degree required
+        // This helper function function takes a fractional output value, and returns
+        // num / denom * fractional_output_value, rounded up
+        fn division_helper(fractional_output_value: u64, num: u128, denom: u128) -> u64 {
+            let num_128 = fractional_output_value as u128 * num;
+            // Divide by fill_fraction_denom, rounding up, and truncate to u64
+            ((num_128 + (denom - 1)) / denom) as u64
+        }
```

Comment:
:thought_balloon: This logic seems important to be correct and ensure it stays correct, however the builder tests seems to be focusing on values that easily divide and would still work if someone got rid of this and did ((output_value * numerator) / denominator)

Reply:
good point, I'll make sure I add a test that hits the boundary condition